### ISSUE TITLE
Cotainr fails to build container when /tmp runs out of space

### ISSUE THREAD

**Issue by Chroxvi** [DeiC-HPC/cotainr#51]
If you try to build a container using cotainr, e.g. `cotainr build lumi_pytorch_rocm_demo.sif --system=lumi-g --conda-env py311_rocm542_pytorch.yml`, on a system which does not have sufficient space on /tmp to store the entire container, you will encounter an error like:

```bash
...
INFO:    Creating sandbox directory...
INFO:    Build complete: /tmp/tmpobd2znt4/singularity_sandbox
Pip subprocess error:
ERROR: Could not install packages due to an OSError: [Errno 28] No space left on device
...
```

Cotainr does not provide a CLI option, environment variable, or similar for changing the location of the temporary sandbox directory, created during the build phase, to another location than /tmp. This might be a problem if:

- You are building a very large container that does not fit /tmp on your system. In that case you can't use cotainr.
- You are building containers on a shared system, e.g. a shared HPC-system login-node, where other people might coincidentally use /tmp. In that case you might experience occasional build failures.

**Comment by Chroxvi** [comment]
If this happens on a login-node on the LUMI supercomputer, as a workaround, you can try to build your container on another node - either another login node or a compute node. To use a LUMI-C compute node, you can submit a job using `srun`, e.g. something along the lines of: `srun --account=project_<your_project_id> --time=00:15:00 --mem=64G --cpus-per-task=32 --partition=small cotainr build lumi_pytorch_rocm_demo.sif --system=lumi-g --conda-env py311_rocm542_pytorch.yml`. Note that you have to request *enough* memory, since /tmp is mounted as an in-memory filesystem on LUMI compute nodes. Also note that building on a compute node will consume CPUh/GPUh resources from your LUMI project.

**Comment by Gabrielgerez** [comment]
Hi I am having this issue with a different cluster trying to build a container with only the pip package tensorflow. 
How does cotainr decide where to build the singularity sandbox? 
Is there a way to change this?

**Comment by Chroxvi** [comment]
TL;DR: You should be able to set the directory used for building the singularity sandbox by setting either of the environment variables `TMPDIR`, `TEMP` or `TMP`.

Cotainr uses a standard Python `tempfile.TemporaryDirectory` for the location of the temporary sandbox directory: https://github.com/DeiC-HPC/cotainr/blob/1ae1b0d8aa885e13e3b17450ba8e69fcf9e228f2/cotainr/container.py#L90-L93

Looking at the [documentation for the `tempfile.gettempdir`](https://docs.python.org/3/library/tempfile.html#tempfile.gettempdir), it says:

```
Return the name of the directory used for temporary files. This defines the default value for the dir argument to all functions in this module.

Python searches a standard list of directories to find one which the calling user can create files in. The list is:
1. The directory named by the TMPDIR environment variable.
2. The directory named by the TEMP environment variable.
3. The directory named by the TMP environment variable.
4. A platform-specific location:
  - On Windows, the directories C:\TEMP, C:\TMP, \TEMP, and \TMP, in that order.
  - On all other platforms, the directories /tmp, /var/tmp, and /usr/tmp, in that order.
5. As a last resort, the current working directory.
```
The above rules should apply to `tempfile.TemporaryDirectory`, and thus also be used for determining the location of the temporary directory used for the singularity sandbox that cotainr creates.

**Comment by Gabrielgerez** [comment]
This seems to have removed the issue with space, thanks!
Though the build crashes now on:
```
SingularitySandbox/CondaInstall.err:-: EnvironmentFileNotFound: '/path/to/my/tmpdir/tmpo33zmhzi/singularity_sandbox/environment.yml' file not found 
```

From these lines the environment file should be getting copied correctly, maybe I am doing something wrong:
https://github.com/DeiC-HPC/cotainr/blob/1ae1b0d8aa885e13e3b17450ba8e69fcf9e228f2/cotainr/cli.py#L227-L232

the sequence of commands I ran were:
```
export TMP=/path/to/my/tmpdir/
cotainr build my_conda_env_container.sif --base-image=docker://ubuntu:22.04 --conda-env=environment.yml -v
```
where the `environment.yml` is:
```
channels:
  - conda-forge
dependencies:
  - python
  - pip
```
all files and commands are in the same `/path/to/my/tmpdir`
This still happens when also trying to install `tensorflow`, this is a minimal example which still does this.

**Comment by Chroxvi** [comment]
@Gabrielgerez Thanks for providing a minimal example of this problem! I can reproduce the problem.

This seems to be an issue with the way we manipulate files when running commands inside the container. We are basically trying to install the conda environment referencing the absolute path on your file system, i.e. `/path/to/my/tmpdir/tmpo33zmhzi/singularity_sandbox/environment.yml`, instead of the relative path when running the command inside the container, i.e. `/environment.yml`. Using the absolute path incidentally happens to work out when using `/tmp` for temporary storage, [as `/tmp` from the host is usually automatically bind mounted when running the Singularity container](https://docs.sylabs.io/guides/3.0/user-guide/bind_paths_and_mounts.html#system-defined-bind-paths)...

I think we need to rework our abstractions around manipulating files within the container, as our current approach seems to cause a lot of subtle bugs. I will discuss this with the rest of the development team as I consider this a somewhat critical bug.

Unfortunately, I don't think we currently have a workaround for this problem as `cotainr` also lacks a way to handle arbitrary Singularity bind mounts.
